storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/writer/RedisClusterCustomIO.java

```diff
@@ -0,0 +1,294 @@
+            .setErrorMessage(exception.getMessage())
+            .setStackTrace(ExceptionUtils.getStackTrace(exception))
+            .build();
+      }
+
+      private byte[] getKey(FeatureRow featureRow) {
+        FeatureSetSpec featureSetSpec = featureSetSpecs.get(featureRow.getFeatureSet());
+        List<String> entityNames =
+            featureSetSpec.getEntitiesList().stream()
+                .map(EntitySpec::getName)
+                .sorted()
+                .collect(Collectors.toList());
+
+        Map<String, Field> entityFields = new HashMap<>();
+        Builder redisKeyBuilder = RedisKey.newBuilder().setFeatureSet(featureRow.getFeatureSet());
+        for (Field field : featureRow.getFieldsList()) {
+          if (entityNames.contains(field.getName())) {
+            entityFields.putIfAbsent(
+                field.getName(),
+                Field.newBuilder().setName(field.getName()).setValue(field.getValue()).build());
+          }
+        }
+        for (String entityName : entityNames) {
+          redisKeyBuilder.addEntities(entityFields.get(entityName));
+        }
+        return redisKeyBuilder.build().toByteArray();
+      }
+
+      private byte[] getValue(FeatureRow featureRow) {
+        FeatureSetSpec spec = featureSetSpecs.get(featureRow.getFeatureSet());
+
+        List<String> featureNames =
+            spec.getFeaturesList().stream().map(FeatureSpec::getName).collect(Collectors.toList());
+        Map<String, Field> fieldValueOnlyMap =
+            featureRow.getFieldsList().stream()
+                .filter(field -> featureNames.contains(field.getName()))
+                .distinct()
+                .collect(
+                    Collectors.toMap(
+                        Field::getName,
+                        field -> Field.newBuilder().setValue(field.getValue()).build()));
+
+        List<Field> values =
+            featureNames.stream()
+                .sorted()
+                .map(
+                    featureName ->
+                        fieldValueOnlyMap.getOrDefault(
+                            featureName,
+                            Field.newBuilder()
+                                .setValue(ValueProto.Value.getDefaultInstance())
+                                .build()))
+                .collect(Collectors.toList());
+
+        return FeatureRow.newBuilder()
+            .setEventTimestamp(featureRow.getEventTimestamp())
+            .addAllFields(values)
+            .build()
+            .toByteArray();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        FeatureRow featureRow = context.element();
+        featureRows.add(featureRow);
+        if (featureRows.size() >= batchSize) {
+          try {
+            executeBatch();
+            featureRows.forEach(row -> context.output(successfulInsertsTag, row));
+            featureRows.clear();
+          } catch (Exception e) {
+            featureRows.forEach(
```

Comment:
Should we be logging to console or to statsd here if we have failures? or does that happen downstream?

Reply:
It would be done downstream: https://github.com/gojek/feast/blob/master/ingestion/src/main/java/feast/ingestion/ImportJob.java#L166